jlm/mlir/backend/JlmToMlirConverter.cpp

```diff
@@ -0,0 +1,222 @@
+      results.push_back(block.getArgument(output->index()));
+    }
+    else
+    {
+      // The identified node should always exist in the hash map of nodes
+      JLM_ASSERT(nodes.find(outputNode) != nodes.end());
+      results.push_back(nodes[outputNode]);
+    }
+  }
+  return results;
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertNode(const rvsdg::node & node, ::mlir::Block & block)
+{
+  if (auto simpleNode = dynamic_cast<const rvsdg::simple_node *>(&node))
+  {
+    return ConvertSimpleNode(*simpleNode, block);
+  }
+  else if (auto lambda = dynamic_cast<const llvm::lambda::node *>(&node))
+  {
+    return ConvertLambda(*lambda, block);
+  }
+  else
+  {
+    std::string unkown = "Unimplemented structural node: " + node.operation().debug_string();
+    JLM_UNREACHABLE(unkown.c_str());
+  }
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertSimpleNode(const rvsdg::simple_node & node, ::mlir::Block & block)
+{
+  if (auto bitsOp = dynamic_cast<const rvsdg::bitconstant_op *>(&(node.operation())))
+  {
+    auto value = bitsOp->value();
+    auto constOp = Builder_->create<::mlir::arith::ConstantIntOp>(
+        Builder_->getUnknownLoc(),
+        value.to_uint(),
+        value.nbits());
+    block.push_back(constOp);
+
+    return constOp;
+  }
+  else
+  {
+    std::string unkown = "Unimplemented simple node: " + node.operation().debug_string();
+    JLM_UNREACHABLE(unkown.c_str());
+  }
+}
+
+::mlir::Value
+JlmToMlirConverter::ConvertLambda(const llvm::lambda::node & lambdaNode, ::mlir::Block & block)
+{
+  ::llvm::SmallVector<::mlir::Type> arguments;
+  for (size_t i = 0; i < lambdaNode.nfctarguments(); ++i)
+  {
+    arguments.push_back(ConvertType(lambdaNode.fctargument(i)->type()));
+  }
+  ::llvm::ArrayRef argumentsArray(arguments);
+
+  ::llvm::SmallVector<::mlir::Type> results;
+  for (size_t i = 0; i < lambdaNode.nfctresults(); ++i)
+  {
+    results.push_back(ConvertType(lambdaNode.fctresult(i)->type()));
+  }
+  ::llvm::ArrayRef resultsArray(results);
+
+  ::llvm::SmallVector<::mlir::Type> lambdaRef;
+  auto refType = Builder_->getType<::mlir::rvsdg::LambdaRefType>(argumentsArray, resultsArray);
+  lambdaRef.push_back(refType);
+
+  ::llvm::SmallVector<::mlir::Value> inputs;
+  // TODO
+  // Populate the inputs
+
+  // Add function attributes, e.g., the function name and linkage
+  ::llvm::SmallVector<::mlir::NamedAttribute> attributes;
+  auto symbolName = Builder_->getNamedAttr(
+      Builder_->getStringAttr("sym_name"),
+      Builder_->getStringAttr(lambdaNode.name()));
+  attributes.push_back(symbolName);
+  auto linkage = Builder_->getNamedAttr(
+      Builder_->getStringAttr("linkage"),
+      Builder_->getStringAttr(jlm::llvm::linkage_to_string(lambdaNode.linkage())));
+  attributes.push_back(linkage);
+  ::llvm::ArrayRef<::mlir::NamedAttribute> attributesRef(attributes);
+
+  auto lambda = Builder_->create<::mlir::rvsdg::LambdaNode>(
+      Builder_->getUnknownLoc(),
+      lambdaRef,
```

Comment:
Same here. Could you inline the lambdaRef?

Reply:
lambdaRef isn't that easy to inline, but the arrayRef has been inlined, similar as above.